pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -1386,10 +1386,15 @@ func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleStat
 		claim = claim.DeepCopy()
 		claim.Status.DriverName = driverName
 		claim.Status.Allocation = allocation
+		// This assumed object is not one which was actually stored in
+		// the apiserver yet! If we receive an update, the assumed
+		// object will get overwritten. The claimAssumeCache handles
+		// this case for us by copying the allocation into the updated
+		// object to produce a newer assumed object.
```

Comment:
@alculquicondor: this is from a new commit that I added at the end to address a race that started to pop up in E2E test runs randomly after splitting the finalizer update out of the claim status update.

What happened is that the finalizer update triggered a claim update which then overwrote this modified claim here, causing the next pod scheduling to reuse the same resource instance because it looked like it was unused (finalizer set, allocation result not).

My fix is to fork the assume cache (temporarily - I think this can be generalized again) and add this:
```patch
***************
*** 173,178 ****
--- 172,178 ----
  	c.rwMutex.Lock()
  	defer c.rwMutex.Unlock()
  
+ 	latestObj := obj
  	if objInfo, _ := c.getObjInfo(name); objInfo != nil {
  		newVersion, err := c.getObjVersion(name, obj)
  		if err != nil {
***************
*** 192,200 ****
  			c.logger.V(10).Info("Skip adding object to assume cache because version is not newer than storedVersion", "description", c.description, "cacheKey", name, "newVersion", newVersion, "storedVersion", storedVersion)
  			return
  		}
  	}
  
! 	objInfo := &objInfo{name: name, latestObj: obj, apiObj: obj}
  	if err = c.store.Update(objInfo); err != nil {
  		c.logger.Info("Error occurred while updating stored object", "err", err)
  	} else {
--- 192,213 ----
  			c.logger.V(10).Info("Skip adding object to assume cache because version is not newer than storedVersion", "description", c.description, "cacheKey", name, "newVersion", newVersion, "storedVersion", storedVersion)
  			return
  		}
+ 
+ 		// If we we have an assumed object, then it always has an
+ 		// allocation. If the new object doesn't, then we have to keep
+ 		// overriding allocation because the scheduler plugin is still
+ 		// in the process of writing out that change.  Eventually, the
+ 		// plugin will store the actual written claim or restore the
+ 		// original claim.
+ 		if claim := obj.(*resourcev1alpha2.ResourceClaim); claim.Status.Allocation == nil {
+ 			claim := claim.DeepCopy()
+ 			claim.Status.DriverName = objInfo.latestObj.(*resourcev1alpha2.ResourceClaim).Status.DriverName
+ 			claim.Status.Allocation = objInfo.latestObj.(*resourcev1alpha2.ResourceClaim).Status.Allocation
+ 			latestObj = claim
+ 		}
  	}
  
! 	objInfo := &objInfo{name: name, latestObj: latestObj, apiObj: obj}
  	if err = c.store.Update(objInfo); err != nil {
  		c.logger.Info("Error occurred while updating stored object", "err", err)
  	} else {
```

This could already happen even with the atomic status update, it just turned from "extremly unlikely" to "happens quite a bit".

Reply:
Hmm, this is causing problems elsewhere. I think what I really need to do is override the allocation of in-flight claims via the `inFlightAllocations`. Stay tuned...

Reply:
Should be done now, I pushed the update. I'm still keeping new commits at the end, so you have a chance to see what changed since before the rebase.

I'm also including further E2E test coverage enhancements. Unit tests are great to prevent regressions, but if the code is wrong, so are often also the unit tests. Nothing ruins a day better than unexpected race conditions in an E2E test... :sweat_smile: